duckduckgo/duckduckgo-privacy-extension
# Refactor the Click to Load declarativeNetRequest rule logic

The "Click to Load" feature blocks some embedded third-party content,
while giving users the option to click to load it again. To get that
working with Chrome MV3, we have the necessary allowing
declarativeNetRequest rules ready and add those for tabs where the
blocking shouldn't happen. There are quite a few things to consider,
like if the content is first-party, if the user has allowlisted the
website, if the feature is enabled for the current domain etc. Let's
refactor the declarativeNetRequest rule generation logic to handle
more of those edge-cases.



**Reviewer:** @jdorweiler 
**CC:** @ladamski 

## Steps to test this PR:
For MV3 builds:
1. Ensure YouTube videos aren't blocked and placeholders aren't shown on [our YouTube test page](https://privacy-test-pages.glitch.me/privacy-protections/youtube-click-to-load/).
2. Ensure Facebook content is blocked correctly and can be unblocked as expected on [our Facebook test page](https://privacy-test-pages.glitch.me/privacy-protections/click-to-load/).
3. Ensure that Facebook login works OK (e.g. for https://www.yelp.co.uk/login )
4. Check that when an exception for the `clickToPlay` feature is added for the `privacy-test-pages.glitch.me` domain in the extension config, Facebook requests are no longer blocked and placeholders are no longer displayed for [our Facebook test page](https://privacy-test-pages.glitch.me/privacy-protections/click-to-load/)
5. Likewise, test that when the user allowlists the test page the content loads.
6. Likewise, test that when the `clickToPlay` feature is entirely disabled the Facebook and YouTube content on our test pages all load.

## Automated tests:

###### Reviewer Checklist:


###### PR Author Checklist:
